dcicutils/qa_utils.py

```diff
@@ -412,25 +424,248 @@ def __exit__(self, exc_type, exc_val, exc_tb):
+        super().__init__(size=size, empty=empty)
+        self.byte_string = content_to_copy
+
+    def copied_content(self):
+        return self.byte_string
+
+
+class MockBigContent(MockPartableContent):
+
+    def start_cloning(self):
+        return MockBigContent(size=self.size, empty=True, content_id=self._content_id)
+
+    def copied_content(self):
+        return self
+
+
+IntPair = List[int]  # it's a list of ints, but for now we know no way to type hint a list of exactly 2 integers
+
+
+# I'm not sure we need this function. We're doing this a different way. -kmp 15-May-2023
+def add_coverage(coverage: List[IntPair], start: int, end: int):
+    return simplify_coverage(coverage + [[start, end]])
+
+
+def simplify_coverage(coverage: Iterable[IntPair]) -> List[IntPair]:
+    current = [0, 0]
+    result = []
+    for item in sorted(coverage):
+        [s, e] = item
+        if e < s:
+            raise ValueError(f"Consistency problem: {item} is out of order.")
+        elif s > current[1]:
+            if current[0] != current[1]:
+                result.append(current)
+            current = [s, e]
+        elif e > current[1]:
+            current[1] = e
+    if current[0] != current[1]:
+        result.append(current)
+    return result
+
+
+def is_abstract_content(content):
+    return isinstance(content, MockAbstractContent)
 
 
 class MockFileSystem:
     """Extremely low-tech mock file system."""
 
     def __init__(self, files=None, default_encoding='utf-8', auto_mirror_files_for_read=False, do_not_auto_mirror=()):
+        files = files or {}
         self.default_encoding = default_encoding
         # Setting this dynamically will make things inconsistent
         self._auto_mirror_files_for_read = auto_mirror_files_for_read
         self._do_not_auto_mirror = set(do_not_auto_mirror or [])
-        self.files = {filename: content.encode(default_encoding) for filename, content in (files or {}).items()}
-        for filename in self.files:
+        # self.files = {filename: content.encode(default_encoding) for filename, content in (files or {}).items()}
+        for filename in files:
             self._do_not_mirror(filename)
+        self.files = {}
+        for filename, content in files.items():
+            self.set_file_content_for_testing(filename, content.encode(default_encoding))
 
     IO_OPEN = staticmethod(io.open)
     OS_PATH_EXISTS = staticmethod(os.path.exists)
     OS_REMOVE = staticmethod(os.remove)
 
+    def assert_file_count(self, n):
+        assert len(self.files) == n
+
+    def _set_file_content_for_testing(self, filename, content):
+        # This is subprimitive to set_file_content_for_testing or others that need to do auxiliary actions as well.
+        self.files[filename] = content
+
+    def set_file_content_for_testing(self, filename, content):
+        # We might at some future point want to consider whether callers of this function should
+        # see auto-mirroring or versioning.
```

Comment:
These types of comments seem to be a mix of docstring vs. inline comments, may be worth generating docstrings for all of these and adding the notes where appropriate